File: jtaf-web/src/main/java/ch/jtaf/report/Diplomas.java
package ch.jtaf.report;

import ch.jtaf.entity.Athlete;
import ch.jtaf.entity.Category;
import ch.jtaf.i18n.I18n;
import ch.jtaf.vo.CompetitionRankingCategoryVO;
import ch.jtaf.vo.CompetitionRankingVO;
import com.itextpdf.text.*;
import com.itextpdf.text.pdf.PdfPCell;
import com.itextpdf.text.pdf.PdfPTable;
import com.itextpdf.text.pdf.PdfWriter;
import org.jboss.logging.Logger;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Locale;

public class Diplomas extends AbstractReport {

    private static final SimpleDateFormat SDF = new SimpleDateFormat("d. MMMMM yyyy");
    private static final float ATHLETE_FONT_SIZE = 12f;

    private Document document;
    private final byte[] logo;
    private final CompetitionRankingVO ranking;

    public Diplomas(CompetitionRankingVO ranking, byte[] logo, Locale locale) {
        super(locale);
        this.ranking = ranking;
        this.logo = logo;
    }

    public byte[] create() {
        try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
            document = new Document(PageSize.A5, cmToPixel(1.5f), cmToPixel(1.5f), cmToPixel(1f), cmToPixel(1.5f));
            PdfWriter pdfWriter = PdfWriter.getInstance(document, baos);
            document.open();
            boolean first = true;
            for (CompetitionRankingCategoryVO cat : ranking.getCategories()) {
                int rank = 1;
                for (Athlete athlete : cat.getAthletes()) {
                    if (!first) {
                        document.newPage();
                    }
                    createTitle();
                    createLogo();
                    createCompetitionInfo();
                    createAthleteInfo(rank, athlete, cat.getCategory());
                    first = false;
                    rank++;
                }
            }
            document.close();
            pdfWriter.flush();
            return baos.toByteArray();
        } catch (DocumentException | IOException e) {
            Logger.getLogger(Diplomas.class).error(e.getMessage(), e);
            return new byte[0];
        }
    }

    private void createLogo() throws DocumentException, IOException {
        if (logo != null) {
            Image image = Image.getInstance(logo);
            image.scaleToFit(cmToPixel(11f), cmToPixel(11f));
            image.setAbsolutePosition((cmToPixel(14.85f) - image.getScaledWidth()) / 2,
                    (cmToPixel(11f) - image.getScaledHeight()) / 2 + cmToPixel(5.5f));
            document.add(image);
        }
    }

    private void createAthleteInfo(int rank, Athlete athlete, Category category) throws DocumentException {
        PdfPTable table = new PdfPTable(new float[]{2f, 10f, 10f, 3f, 2f});
        table.setWidthPercentage(100);
        table.setSpacingBefore(cmToPixel(1.5f));

        addCell(table, rank + ".", ATHLETE_FONT_SIZE);
        addCell(table, athlete.getLastName(), ATHLETE_FONT_SIZE);
        addCell(table, athlete.getFirstName(), ATHLETE_FONT_SIZE);
        addCell(table, athlete.getYear() + "", ATHLETE_FONT_SIZE);
        addCell(table, category.getAbbreviation(), ATHLETE_FONT_SIZE);

        document.add(table);
    }

    private void createTitle() throws DocumentException {
        PdfPTable table = new PdfPTable(1);
        table.setWidthPercentage(100);

        PdfPCell cell = new PdfPCell(
                new Phrase(I18n.getInstance().getString(locale, "Diploma"),
                        FontFactory.getFont(FontFactory.HELVETICA, 60f)));
        cell.setBorder(0);
        cell.setHorizontalAlignment(PdfPCell.ALIGN_CENTER);

        table.addCell(cell);
        document.add(table);
    }

    private void createCompetitionInfo() throws DocumentException {
        PdfPTable table = new PdfPTable(1);
        table.setWidthPercentage(100);
        table.setSpacingBefore(cmToPixel(12f));

        PdfPCell cell = new PdfPCell(
                new Phrase(ranking.getCompetition().getName(),
                        FontFactory.getFont(FontFactory.HELVETICA, 25f)));
        cell.setBorder(0);
        cell.setHorizontalAlignment(PdfPCell.ALIGN_CENTER);
        table.addCell(cell);
        
        cell = new PdfPCell(
                new Phrase(SDF.format(ranking.getCompetition().getCompetitionDate()),
                        FontFactory.getFont(FontFactory.HELVETICA, 25f)));
        cell.setBorder(0);
        cell.setHorizontalAlignment(PdfPCell.ALIGN_CENTER);
        table.addCell(cell);

        document.add(table);
    }
}

File: src/main/kotlin/ch/jtaf/control/reporting/report/Diplomas.kt
package ch.jtaf.control.reporting.report

import ch.jtaf.control.reporting.data.CompetitionRankingData
import ch.jtaf.entity.AthleteWithResultsDTO
import ch.jtaf.entity.Category
import com.itextpdf.text.*
import com.itextpdf.text.pdf.PdfPCell
import com.itextpdf.text.pdf.PdfPTable
import com.itextpdf.text.pdf.PdfWriter
import org.slf4j.LoggerFactory
import java.io.ByteArrayOutputStream
import java.text.SimpleDateFormat
import java.util.*

class Diplomas(private val ranking: CompetitionRankingData, private val logo: ByteArray?, locale: Locale) : AbstractReport(locale) {

    private val logger = LoggerFactory.getLogger(Diplomas::class.java)

    private val simpleDateFormat = SimpleDateFormat("d. MMMMM yyyy")
    private val athleteFontSize = 12f

    private var document: Document? = null

    @Throws(DocumentException::class)
    fun create(): ByteArray {
        try {
            ByteArrayOutputStream().use { baos ->
                document = Document(PageSize.A5, cmToPixel(1.5f), cmToPixel(1.5f), cmToPixel(1f), cmToPixel(1.5f))
                val pdfWriter = PdfWriter.getInstance(document!!, baos)
                document!!.open()
                var first = true
                for (cat in ranking.categories) {
                    var rank = 1
                    cat.getAthletesSortedByPointsDesc().forEach {
                        if (!first) {
                            document!!.newPage()
                        }
                        createTitle()
                        createLogo()
                        createCompetitionInfo()
                        createAthleteInfo(rank, it, cat.category)
                        first = false
                        rank++
                    }
                }
                document!!.close()
                pdfWriter.flush()
                return baos.toByteArray()
            }
        } catch (e: Exception) {
            logger.error(e.message, e)
            return ByteArray(0)
        }

    }

    private fun createLogo() {
        if (logo != null) {
            val image = Image.getInstance(logo)
            image.scaleToFit(cmToPixel(11f), cmToPixel(11f))
            image.setAbsolutePosition((cmToPixel(14.85f) - image.scaledWidth) / 2,
                    (cmToPixel(11f) - image.scaledHeight) / 2 + cmToPixel(5.5f))
            document!!.add(image)
        }
    }

    private fun createAthleteInfo(rank: Int, athlete: AthleteWithResultsDTO, category: Category) {
        val table = PdfPTable(floatArrayOf(2f, 10f, 10f, 3f, 2f))
        table.widthPercentage = 100f
        table.spacingBefore = cmToPixel(1.5f)

        addCell(table, rank.toString() + ".", athleteFontSize)
        addCell(table, athlete.athlete.lastName, athleteFontSize)
        addCell(table, athlete.athlete.firstName, athleteFontSize)
        addCell(table, athlete.athlete.yearOfBirth.toString(), athleteFontSize)
        addCell(table, category.abbreviation, athleteFontSize)

        document!!.add(table)
    }

    private fun createTitle() {
        val table = PdfPTable(1)
        table.widthPercentage = 100f

        val cell = PdfPCell(Phrase("Diploma", FontFactory.getFont(FontFactory.HELVETICA, 60f)))
        cell.border = 0
        cell.horizontalAlignment = PdfPCell.ALIGN_CENTER

        table.addCell(cell)
        document!!.add(table)
    }

    private fun createCompetitionInfo() {
        val table = PdfPTable(1)
        table.widthPercentage = 100f
        table.spacingBefore = cmToPixel(12f)

        var cell = PdfPCell(Phrase(ranking.competition.name, FontFactory.getFont(FontFactory.HELVETICA, 25f)))
        cell.border = 0
        cell.horizontalAlignment = PdfPCell.ALIGN_CENTER
        table.addCell(cell)

        cell = PdfPCell(Phrase(simpleDateFormat.format(ranking.competition.competitionDate), FontFactory.getFont(FontFactory.HELVETICA, 25f)))
        cell.border = 0
        cell.horizontalAlignment = PdfPCell.ALIGN_CENTER
        table.addCell(cell)

        document!!.add(table)
    }

}
